jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,369 @@
+                maxRequests = ProcessorUtils.availableProcessors();
+            setMaxRequests(maxRequests);
+        }
+        semaphore = new Semaphore(maxRequests);
+
+        int maxPriority = getMaxPriority();
+        int capacity = maxPriority + 1;
+        queues = new ArrayList<>(capacity);
+        for (int i = 0; i < capacity; ++i)
+        {
+            queues.add(new ConcurrentLinkedQueue<>());
+        }
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} initialized maxRequests={} maxPriority={}", this, maxRequests, maxPriority);
+
+        super.doStart();
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        super.doStop();
+        removeBean(timeouts);
+        timeouts.destroy();
+    }
+
+    @Override
+    public boolean handle(Request request, Response response, Callback callback) throws Exception
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} handling {}", this, request);
+        return process(request, response, callback);
+    }
+
+    private boolean process(Request request, Response response, Callback callback) throws Exception
+    {
+        if (semaphore.tryAcquire())
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} forwarding {}", this, request);
+            request.addHttpStreamWrapper(stream -> new Resumer(stream, request));
+            return super.handle(request, response, callback);
+        }
+        else
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} suspending {}", this, request);
+            int priority = getPriority(request);
+            priority = Math.min(Math.max(0, priority), getMaxPriority());
+            Entry entry = new Entry(request, response, callback, priority);
+            queues.get(priority).offer(entry);
+            timeouts.schedule(entry);
+            return true;
+        }
+    }
+
+    /**
+     * <p>Returns the priority of the given suspended request.</p>
+     *
+     * @param request the suspended request to compute the priority for
+     * @return the priority of the given suspended request
+     * @see #setMaxPriority(int)
+     */
+    protected int getPriority(Request request)
+    {
+        return 0;
+    }
+
+    private void processAndComplete(Request request, Response response, Callback callback)
+    {
+        try
+        {
+            boolean handled = process(request, response, callback);
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} handled={} {}", this, handled, request);
+            if (!handled)
+                Response.writeError(request, response, callback, HttpStatus.NOT_FOUND_404);
+        }
+        catch (Throwable x)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} failed {}", this, request, x);
+            Response.writeError(request, response, callback, x);
+        }
+    }
+
+    private void resume()
+    {
+        semaphore.release();
```

Comment:
since it is this thread that is going to do the `processAndComplete` do we really need to release the semaphore and then look for an entry?  isn't there a risk that a newly arrived job will steal the semaphore even if it has a lower priority, before we acquire it again.

Can't we just keep the semaphore and go on to process the highest priority entry.  We should only release if we don't find such an entry.

Reply:
Good point, I'll do that.